keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -0,0 +1,578 @@
+
+Protobuf support. Apply does not support protobuf, and it will not be added with
+this enhancement.
+
+## Proposal
+
+`Apply` functions should be included in the typed clients generated for
+client-go and should accept the apply configuration using a strongly typed
+representation, which will need to be generated for this purpose.
+
+### Risks and Mitigations
+
+#### Poor adoption
+
+Risk: Developers adoption is poor, either because the aesthetics/ergonomics are
+not to their liking or the functinality is insufficient to do what they need to
+do with it. This could lead to (a) poor server side apply adoption, and/or (b)
+developers building alternate solutions.
+
+Mitigation: We are working with the kubebuilder community to
+get hands on feedback from developers to guide our design decisions around
+aesthetics/ergonomics with a goal of having both client-go and kubebuilder
+take an aligned approach to adding apply to clients in a typesafe way.
+
+## Design Details
+
+### Apply functions
+
+The client-go typed clients will be extended to include Apply functions, e.g.:
+
+```go
+func (c *deployments) Apply(ctx Context, fieldManager string, deployment *appsv1apply.Deployment, opts metav1.ApplyOptions, subresources ...string) (*Deployment, error)
+```
+
+`ApplyOptions` will be added to metav1 even though `PatchOptions` will continue
+to be used over the wire. This will make it obvious in the Apply function
+signature that `fieldManager` is required.
+
+```go
+type ApplyOptions struct {
+  DryRun []string `json:"dryRun,omitempty" protobuf:"bytes,1,rep,name=dryRun"`
+  Force *bool `json:"force,omitempty" protobuf:"varint,2,opt,name=force"`
+}
+
+func (ApplyOptions) ToPatchOptions(fieldManager string) PatchOptions
+```
+
+<<[UNRESOLVED @jpbetz ]>> 
+Should the Force field also be made a required field like fieldManager is?
+@lavalamp pointed out that user things typically want false but
+controller-things typically want true, so false isn't a great default.
+
+Also, can we structure this so that we don't have so many positional params?
+Moving two params from options to the top level positional params results in a
+longer function signature. Alternative proposed by @lavalamp was to do something
+more like this (ordering or parameters and function names TBD):
+
+```go
+client.Deployments().BeginApply("field-manager-name", options).Body(
+  appsv1apply.Deployment().
+    SetObjectMeta(&metav1apply.ObjectMeta()). ...
+).Apply(ctx)
+```
+
+<<[/UNRESOLVED]>>
+
+Apply will combine the `fieldManager` argument with `ApplyOptions` to create the
+`PatchOptions`.
+
+Each apply call will be required to provide a fieldmanager name. We will not
+provide a a way for the fieldmanager name to be set for the entire
+clientset. The main reason for this is that if a client has multiple code paths
+where it makes apply requests to the same object, but with different field sets,
+they must use different field manager names. If they use the same field manager
```

Comment:
could this be satisfied with a "normal" helper for construction, `NewApplyOptions(fieldManager, force)` and keep the symmetric client methods?

Reply:
This is a good alternative. I hadn't thought of it. Adding it to the `<<UNRESOLVED` I've got for this whole field manager topic.